### DISCUSSION TITLE
Finished status of nested states

### DISCUSSION THREAD

**Discussion by oyanyev** [nsk90/kstatemachine#67] [General]
I would like to discuss the way states with child states are assigned finished status.
```
private suspend fun handleStateEntry(state: InternalState, transitionParams: TransitionParams<*>) {
        val finished = when (childMode) {
            **EXCLUSIVE -> state is IFinalState**
            PARALLEL -> states.all { it.isFinished }
        }
...
```
Above is code the finishes parent state immediately when child final state is entered. 
I would prefer outer state to finish ONLY when any child final state is finished (recursively). Until that I would like to receive common for all nested states events in parent state.
So the lifecycle of parent event should not be shorter than the lifecycle of its nested events.

P.S. I am talking about EXCLUSIVE mode

**Comment by nsk90** [comment]
Hi, `IFinalState` is a marker that tells that entry to marked state finishes its parent. 

As I know this is common behaviour for state machine libraries in different languages. 
As many of them follow this two docs same as kstatemachine:
https://www.wisdom.weizmann.ac.il/~dharel/SCANNED.PAPERS/Statecharts.pdf
http://www.w3.org/TR/scxml/

If you need some other notification logics it should be possible to implement it in your application code. 
If you face some problems with such implementation (for example kstatemachine hides some required information) I am ready to discuss and implement some solution. But I don't think that changing current behaviour is an option.

**Reply by oyanyev** [reply]
Isn't that strange when parent State is finished prior to ist child? 
I'm implementing quite complex SM. And do not want to handle common events in each child event.
But if I handle it in parent and parent is finished, then that does not work.

**Reply by oyanyev** [reply]
```
When a state machine enters a <final> substate of a compound state, the parent state remains active, but the event "done.state.id" is generated, where id is the state id of the parent state. This event can trigger a transition in any ancestor state (including the parent).
```
As I see from implementation, parent state remains active, but that does not mean it can process events
```\
    private suspend fun <E : Event> doProcessEvent(eventAndArgument: EventAndArgument<E>): Boolean {
        val (event, argument) = eventAndArgument
        if (isFinished) {
            log { "$this is finished, skipping event ${event::class.simpleName}, with argument $argument" }
            return false
        }
```

**Reply by nsk90** [reply]
> Isn't that strange when parent State is finished prior to ist child? I'm implementing quite complex SM. And do not want to handle common events in each child event. But if I handle it in parent and parent is finished, then that does not work.

It is not strange because not all states can finish (are finishable). So even with recursive finishing strategy any finished state may have children that are not finished (as they are not finishable at all).

**Reply by nsk90** [reply]
Maybe you can provide same little sample so it will be easier to discuss current and expected behaviour?
